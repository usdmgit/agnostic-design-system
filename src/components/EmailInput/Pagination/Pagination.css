@import '../../styles/theme/index.css';

.pagination-list {
  list-style: var(--pagination-list-style, none);
  display: flex;
  gap: var(--pagination-flex-gap, var(--spacing-horizontal-xs));
  flex-direction: var(--pagination-flex-direction, row);
}