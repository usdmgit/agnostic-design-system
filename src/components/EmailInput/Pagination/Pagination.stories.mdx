# Pagination Component

## Type: Simple

## Pagination Props

|           Prop           |   Type    |                                                           Description                                                            | Default | Mandatory |
| :----------------------: | :-------: | :------------------------------------------------------------------------------------------------------------------------------: | :-----: | :-------: |
|        totalPages        |  number   |                                     Specifies the total number of pages we need to paginate.                                     |   10    |   true    |
|   numberOfPagesToShow    | function  | Specifies how many pages we should display on the screen, if 0, no page numbers will be displayed between prev and next buttons. |    3    |   true    |
|       currentPage        |  number   |                          Specifies what is the currrent/active page the consumer is showing/displaying.                          |    1    |   true    |
|       onSelectPage       |  number   |                                     Specifies what should happen when a user selects a page.                                     |    -    |   true    |
|     buttonNextLabel      |  string   |                                     Specifies the text we should display on the next button.                                     | 'Next'  |   false   |
|      buttonNextIcon      | ReactNode |                                        Specifies the icon to be used for the next button.                                        |    -    |   false   |
|  buttonNextIconPosition  | Position  |                                 Specifies the position (left/right) the next button should have.                                 | 'right' |   false   |
|     buttonPrevLabel      |  string   |                                     Specifies the text we should display on the prev button.                                     | 'Prev'  |   false   |
|      buttonPrevIcon      | ReactNode |                                      Specifies the icon to be used for the previous button.                                      |    -    |   false   |
|  buttonPrevIconPosition  | Position  |                                 Specifies the position (left/right) the next button should have.                                 | 'left'  |   false   |
| buttonVariablesClassName |  string   |                             Allow the customization/CSS styling of the button for the previous page.                             |    -    |   false   |

## Getting Started

```javascript
import React, { useState } from 'react';
import { Pagination } from '@codelittinc/agnostic-design-system';

const Pagination = args => {
  const [page, setPage] = useState(args.currentPage);

  return <Pagination {...args} currentPage={page} onSelectPage={e => setPage(e)} />;
};

export default Pagination;
```

## CSS Variables Available

| Property |   Attribute    | State |
| :------: | :------------: | :---: |
|          |   list-style   |       |
|          |    display     |       |
|          |      gap       |       |
|          | flex-direction |       |
