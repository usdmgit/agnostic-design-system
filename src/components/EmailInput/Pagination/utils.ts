export const getSelectedPage = (currentPage: number, totalPages: number) =>
  Math.max(Math.min(currentPage, totalPages), 1);
