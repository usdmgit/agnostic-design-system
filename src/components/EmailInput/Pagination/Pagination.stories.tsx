import React, { useState } from 'react';
import Pagination from '@/components/Pagination';
import { Meta } from '@storybook/react';
import mdx from './Pagination.stories.mdx';

export default {
  title: 'Components/Pagination',
  component: Pagination,
  parameters: {
    docs: {
      page: mdx
    }
  }
} as Meta;

const Template = args => {
  const [page, setPage] = useState(args.currentPage);

  return <Pagination {...args} currentPage={page} onSelectPage={e => setPage(e)} />;
};

export const Default = Template.bind({});
Default.args = {
  totalPages: 10,
  numberOfPagesToShow: 3,
  currentPage: 1
};

export const CustomLabelAndIcon = Template.bind({});
CustomLabelAndIcon.args = {
  totalPages: 10,
  numberOfPagesToShow: 3,
  currentPage: 1,
  buttonNextLabel: 'Next',
  buttonNextIcon: <span> &gt; </span>,
  buttonPrevLabel: 'Previous',
  buttonPrevIcon: <span> &lt; </span>
};
