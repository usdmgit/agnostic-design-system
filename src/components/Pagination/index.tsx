import React from 'react';
// eslint-disable-next-line no-unused-vars
import Button, { Position } from '@/components/Button';
import classNames from 'classnames';
import { getSelectedPage } from './utils';

import styles from '@/components/Pagination/Pagination.css';
export interface Props {
  totalPages: number;
  numberOfPagesToShow: number;
  currentPage: number;
  onSelectPage: (item: number) => any;
  buttonNextLabel?: string;
  buttonNextIcon?: React.ReactNode;
  buttonNextIconPosition?: Position;
  buttonPrevLabel?: string;
  buttonPrevIcon?: React.ReactNode;
  buttonPrevIconPosition?: Position;
  buttonVariablesClassName?: string;
  variablesClassName?: string;
}

const Pagination = (props: Props) => {
  const {
    totalPages,
    numberOfPagesToShow,
    currentPage,
    onSelectPage,
    buttonNextLabel,
    buttonNextIcon,
    buttonNextIconPosition,
    buttonPrevLabel,
    buttonPrevIcon,
    buttonPrevIconPosition,
    buttonVariablesClassName,
    variablesClassName
  } = props;

  const selectedPage = getSelectedPage(currentPage, totalPages);
  const nonSelected = Math.floor(numberOfPagesToShow / 2);
  const min = selectedPage - nonSelected > 0 ? selectedPage - nonSelected : 1;
  const pagesLength = numberOfPagesToShow < totalPages ? numberOfPagesToShow : totalPages;

  const pages = Array.from({ length: pagesLength }, (_, i) => {
    if (totalPages < numberOfPagesToShow) {
      return min + i;
    } else if (totalPages < numberOfPagesToShow + selectedPage) {
      return totalPages - numberOfPagesToShow + i + 1;
    } else {
      return min + i;
    }
  });

  return (
    <ul
      className={classNames(styles['pagination-list'], variablesClassName)}
      data-testid='pagination-element'
    >
      <li>
        <Button
          variablesClassName={buttonVariablesClassName}
          withAppendIcon={!!buttonPrevIcon}
          appendIcon={buttonPrevIcon}
          iconPosition={buttonPrevIconPosition}
          size='medium'
          text={buttonPrevLabel}
          disabled={selectedPage === 1}
          onClick={() => onSelectPage(selectedPage - 1)}
        />
      </li>
      {numberOfPagesToShow > 0 &&
        pages.map(item => (
          <li key={item}>
            <Button
              size='medium'
              text={item.toString()}
              category={selectedPage === item ? 'primary' : 'neutral'}
              onClick={() => onSelectPage(item)}
            />
          </li>
        ))}
      <li>
        <Button
          variablesClassName={buttonVariablesClassName}
          withAppendIcon={!!buttonPrevIcon}
          appendIcon={buttonNextIcon}
          iconPosition={buttonNextIconPosition}
          size='medium'
          text={buttonNextLabel}
          disabled={selectedPage === totalPages}
          onClick={() => onSelectPage(selectedPage + 1)}
        />
      </li>
    </ul>
  );
};

export default Pagination;

Pagination.defaultProps = {
  numberOfPagesToShow: 3,
  currentPage: 1,
  buttonNextLabel: 'Next',
  buttonNextIconPosition: 'right',
  buttonPrevLabel: 'Prev',
  buttonPrevIconPosition: 'left'
};
