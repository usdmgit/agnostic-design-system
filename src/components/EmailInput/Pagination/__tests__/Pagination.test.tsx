import React from 'react';
// eslint-disable-next-line no-unused-vars
import Pagination, { Props } from '@/components/Pagination';
import { mount } from 'enzyme';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

const FIRST_PAGE = 1;
const LAST_PAGE = 5;

describe('Pagination', () => {
  it('renders correctly', () => {
    const onSelectPage = jest.fn();
    const wrapper = mount(
      <Pagination
        totalPages={10}
        numberOfPagesToShow={3}
        currentPage={FIRST_PAGE}
        onSelectPage={onSelectPage}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('Number of pages to show', () => {
    it('needs to be smaller than total pages', () => {
      const onSelectPage = jest.fn();
      renderPagination({ onSelectPage, totalPages: 3, numberOfPagesToShow: 5 });
      const paginationList = screen.getByTestId('pagination-element');
      expect(paginationList).toHaveTextContent(/1/);
      expect(paginationList).toHaveTextContent(/2/);
      expect(paginationList).toHaveTextContent(/3/);
      expect(paginationList).not.toHaveTextContent(/0/);
      expect(paginationList).not.toHaveTextContent(/-1/);
    });
  });

  describe('Number of pages', () => {
    it('is established by the user', () => {
      const onSelectPage = jest.fn();
      renderPagination({ onSelectPage });
      const paginationList = screen.getByTestId('pagination-element');
      expect(paginationList).toHaveTextContent(/1/);
      expect(paginationList).toHaveTextContent(/2/);
      expect(paginationList).toHaveTextContent(/3/);
      expect(paginationList).not.toHaveTextContent(/4/);
      expect(paginationList).not.toHaveTextContent(/5/);
    });
  });

  describe('Usage of the prev button', () => {
    it('is disabled if you are on the first page', () => {
      const onSelectPage = jest.fn();
      const utils = renderPagination({ onSelectPage });
      const prevButton = utils.getByText(/Prev/i).closest('button') as HTMLButtonElement;
      expect(prevButton.disabled).toBeTruthy();
    });

    it('is enabled if previous pages exist', () => {
      const onSelectPage = jest.fn();
      const utils = renderPagination({ onSelectPage, currentPage: FIRST_PAGE + 1 });
      const prevButton = utils.getByText(/Prev/i).closest('button') as HTMLButtonElement;
      expect(prevButton.disabled).not.toBeTruthy();
    });

    it('calls onSelectPage prop with the correct page', () => {
      const onSelectPage = jest.fn();
      renderPagination({ onSelectPage, currentPage: FIRST_PAGE + 1 });
      fireEvent.click(screen.getByText(/Prev/));
      expect(onSelectPage).toHaveBeenCalledWith(FIRST_PAGE);
    });
  });

  describe('Usage of the page number buttons', () => {
    it('calls onSelectPage prop with the correct number', () => {
      const onSelectPage = jest.fn();
      renderPagination({ onSelectPage, currentPage: FIRST_PAGE + 2 });
      fireEvent.click(screen.getByText(/4/));
      expect(onSelectPage).toHaveBeenCalledWith(FIRST_PAGE + 3);
    });
  });

  describe('Usage of the next button', () => {
    it('is disabled if you are on the last page', () => {
      const onSelectPage = jest.fn();
      const utils = renderPagination({ onSelectPage, currentPage: LAST_PAGE });
      const nextButton = utils.getByText(/next/i).closest('button') as HTMLButtonElement;
      expect(nextButton.disabled).toBeTruthy();
    });

    it('is enabled if next page exists', () => {
      const onSelectPage = jest.fn();
      const utils = renderPagination({ onSelectPage, currentPage: FIRST_PAGE + 2 });
      const nextButton = utils.getByText(/next/i).closest('button') as HTMLButtonElement;
      expect(nextButton.disabled).not.toBeTruthy();
    });

    it('calls onSelectPage prop with the correct page', () => {
      const onSelectPage = jest.fn();
      renderPagination({ onSelectPage, currentPage: FIRST_PAGE + 1 });
      fireEvent.click(screen.getByText(/Next/));
      expect(onSelectPage).toHaveBeenCalledWith(FIRST_PAGE + 2);
    });
  });
});

function renderPagination(props: Partial<Props> = {}) {
  return render(
    <Pagination
      totalPages={LAST_PAGE}
      numberOfPagesToShow={3}
      currentPage={FIRST_PAGE}
      onSelectPage={e => e}
      {...props}
    />
  );
}
