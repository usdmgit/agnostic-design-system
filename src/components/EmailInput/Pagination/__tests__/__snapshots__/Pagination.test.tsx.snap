// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<Pagination
  buttonNextIconPosition="right"
  buttonNextLabel="Next"
  buttonPrevIconPosition="left"
  buttonPrevLabel="Prev"
  currentPage={1}
  numberOfPagesToShow={3}
  onSelectPage={[MockFunction]}
  totalPages={10}
>
  <ul
    className="pagination-list"
    data-testid="pagination-element"
  >
    <li>
      <ForwardRef
        disabled={true}
        iconPosition="left"
        onClick={[Function]}
        size="medium"
        text="Prev"
        withAppendIcon={false}
      >
        <ForwardRef
          disabled={true}
          iconPosition="left"
          onClick={[Function]}
          size="medium"
          text="Prev"
          withAppendIcon={false}
        >
          <ForwardRef
            className="button button--neutral"
            disabled={true}
            fixed={false}
            iconPosition="left"
            onClick={[Function]}
            size="medium"
            text="Prev"
            withAppendIcon={false}
          >
            <button
              className="button button--neutral button button--medium"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <span
                className="button-text"
              >
                Prev
              </span>
            </button>
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
    </li>
    <li
      key="1"
    >
      <ForwardRef
        category="primary"
        onClick={[Function]}
        size="medium"
        text="1"
      >
        <ForwardRef
          category="primary"
          onClick={[Function]}
          size="medium"
          text="1"
        >
          <ForwardRef
            category="primary"
            className="button button--primary"
            disabled={false}
            fixed={false}
            iconPosition="right"
            onClick={[Function]}
            size="medium"
            text="1"
          >
            <button
              category="primary"
              className="button button--primary button button--medium"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <span
                className="button-text"
              >
                1
              </span>
            </button>
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
    </li>
    <li
      key="2"
    >
      <ForwardRef
        category="neutral"
        onClick={[Function]}
        size="medium"
        text="2"
      >
        <ForwardRef
          category="neutral"
          onClick={[Function]}
          size="medium"
          text="2"
        >
          <ForwardRef
            category="neutral"
            className="button button--neutral"
            disabled={false}
            fixed={false}
            iconPosition="right"
            onClick={[Function]}
            size="medium"
            text="2"
          >
            <button
              category="neutral"
              className="button button--neutral button button--medium"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <span
                className="button-text"
              >
                2
              </span>
            </button>
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
    </li>
    <li
      key="3"
    >
      <ForwardRef
        category="neutral"
        onClick={[Function]}
        size="medium"
        text="3"
      >
        <ForwardRef
          category="neutral"
          onClick={[Function]}
          size="medium"
          text="3"
        >
          <ForwardRef
            category="neutral"
            className="button button--neutral"
            disabled={false}
            fixed={false}
            iconPosition="right"
            onClick={[Function]}
            size="medium"
            text="3"
          >
            <button
              category="neutral"
              className="button button--neutral button button--medium"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <span
                className="button-text"
              >
                3
              </span>
            </button>
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
    </li>
    <li>
      <ForwardRef
        disabled={false}
        iconPosition="right"
        onClick={[Function]}
        size="medium"
        text="Next"
        withAppendIcon={false}
      >
        <ForwardRef
          disabled={false}
          iconPosition="right"
          onClick={[Function]}
          size="medium"
          text="Next"
          withAppendIcon={false}
        >
          <ForwardRef
            className="button button--neutral"
            disabled={false}
            fixed={false}
            iconPosition="right"
            onClick={[Function]}
            size="medium"
            text="Next"
            withAppendIcon={false}
          >
            <button
              className="button button--neutral button button--medium"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <span
                className="button-text"
              >
                Next
              </span>
            </button>
          </ForwardRef>
        </ForwardRef>
      </ForwardRef>
    </li>
  </ul>
</Pagination>
`;
